### PR TITLE
Revise vector operations in SolverIDR to improve performance

### PR DESCRIPTION
**Pull request by kronbichler**

### CODE REVIEW

File: include/deal.II/lac/solver_idr.h [hunk 1]
```diff
@@ -436,43 +429,58 @@ SolverIDR<VectorType>::solve(const MatrixType &        A,
...
+          for (unsigned int i = k + 1; i < s; ++i)
             M(i, k) = Q[i] * G[k];
 
-          // Orthogonalize r to Q0,...,Qk,
-          // update x
+          // Orthogonalize r to Q0,...,Qk, update x
           {
-            double beta = phi(k) / M(k, k);
-            r.add(-1.0 * beta, G[k]);
+            const double beta = phi(k) / M(k, k);
```
Comment (peterrum) [inline include/deal.II/lac/solver_idr.h:482]: `double`?

**Review by peterrum — APPROVED** [review]

File: include/deal.II/lac/solver_idr.h (hunk 1 above)
Comment (kronbichler) [reply include/deal.II/lac/solver_idr.h:482]: Do you want me to use `VectorType::value_type` like in the other PR? I can do that, if you agree with changing the matrix types and all other stuff currently in `double` variables, such as this one here: https://github.com/dealii/dealii/blob/fb2242de546e203f273a4ab674b7b7de9994c84d/include/deal.II/lac/solver_idr.h#L423-L424
Or are you suggesting something else?

Comment (peterrum) [reply include/deal.II/lac/solver_idr.h:482]: > Do you want me to use VectorType::value_type like in the other PR? I can do that, if you agree with changing the matrix types and all other stuff currently in double variables, such as this one here:

Yes, sounds reasonable :+1:

**Review by tjhei — APPROVED** [review]
Very nice. Getting rid of some of the temporaries/reductions might be visible even for situations where the preconditioner is expensive...

**Review by peterrum — APPROVED** [review]

**Pull request merged**